[v5] Input types

davidkpiano opened the pull request:
This PR strongly types `input` values:

```ts
const machine = createMachine({
  types: {} as {
    input: { count: number }
  },
  context: ({ input }) => ({
    value: input.count // strongly-typed
  })
});

interpret(machine, {
  input: { count: 42 } // strongly-typed
}).start();
```

mellson commented on packages/core/src/StateNode.ts:
```diff
@@ -104,7 +104,7 @@ export class StateNode<
    * The root machine node.
    */
-  public machine: StateMachine<TContext, TEvent, any, any>;
+  public machine: StateMachine<TContext, TEvent, any, any, TODO, TODO>;
```
First time I've seen a `<TODO>` generic type, brilliant 😆

Andarist commented on packages/core/test/input.test.ts:
```diff
@@ -94,8 +135,15 @@ describe('input', () => {
+        return input;
+      },
+      entry: ({ context }) => {
+        expect(context.greeting).toBe('hello');
```
```suggestion
        expect(context.greeting).toBe('hello');
        expect(event.input.greeting).toBe('hello');
```

Andarist commented on packages/core/test/input.test.ts:
```diff
@@ -76,8 +112,13 @@ describe('input', () => {
+      },
+      context: ({ input }) => input,
+      entry: ({ context }) => {
+        expect(context.greeting).toBe('hello');
```
```suggestion
        expect(context.greeting).toBe('hello');
        expect(event.input.greeting).toBe('hello');
```

Andarist commented on packages/core/test/input.test.ts:
```diff
@@ -21,29 +31,55 @@ describe('input', () => {
     expect(spy).toHaveBeenCalledWith(42);
   });
 
   it('initial event should have input property', (done) => {
```
this test case no longer tests what it should :wink:

davidkpiano replied:
Yeah, we shouldn't rely on the event having the input data. That should be saved to `context` and read there as a best practice

Andarist replied:
As a best practice - sure, that's probably for the better. As it stands though, this is also part of the init event (creating a nice symmetry with "done" event that contains `output` and we can't easily remove it from there). We should be careful when adjusting tests - we don't want to lose valuable tests and assertions. Even if we'd like to remove `input` from the init event then this whole test should be gone (or adjusted to document that there is only `type` property on the init event).

Andarist replied:
Keeping `input` on the init event has also one pretty nice advantage - it will be easier to keep this information around in the inspector tools etc

Andarist commented on packages/core/src/types.ts:
```diff
@@ -1186,7 +1202,11 @@ export type Mapper<
+> = (args: {
+  context: TContext;
+  event: TEvent;
+  self: ActorRef<TEvent>;
```
this has been added here without any test coverage 😢

davidkpiano replied:
Sorry! I'll add tests

Andarist commented on packages/core/src/utils.ts:
```diff
@@ -231,10 +232,11 @@ export function mapContext<
 ): any {
   if (typeof mapper === 'function') {
-    return mapper({ context, event });
+    return mapper({ context, event, self });
```
`self` is not passed here to the property mapper (`subMapper`) that is called below

davidkpiano commented:
@Andarist I removed the property mapper type from `output` (was never on `input`) and added a dev-time warning:

```
  console.warn
    Dynamically mapping values to individual properties is deprecated. Use a single function that returns the mapped object instead.
    Found object containing properties whose values are possibly mapping functions: 
     - user: ({event}) => event.user
```

davidkpiano commented:
@Andarist Feel free to approve or merge when ready

Andarist commented on packages/core/test/final.test.ts:
```diff
@@ -113,7 +113,7 @@ describe('final states', () => {
               type: 'final',
               output: {
-                secret: () => 'the secret'
+                secret: 'the secret'
```
This test no longer tests what it should (`'should call data expressions on nested final nodes'`). To fix this we could make the whole `output` an expression here

Andarist commented on packages/core/src/utils.ts:
```diff
@@ -239,20 +239,27 @@ export function mapContext<
-      result[key] = subMapper(args);
-    } else {
-      result[key] = subMapper;
+  if (isDevelopment) {
+    if (typeof mapper === 'object') {
+      if (Object.values(mapper).some((val) => typeof val === 'function')) {
```
nit: this could be a single if statement: `a && b && c`
